rclcpp_action/src/server.cpp

```diff
@@ -645,6 +649,37 @@ ServerBase::execute_check_expired_goals()
   }
 }
 
+void ServerBase::setup_expire_goal_timer()
+{
+  std::lock_guard<std::mutex> lock(expire_goal_timers_mutex_);
+  unsigned int timer_id = expire_goal_timers_.empty() ? 1 : expire_goal_timers_.back().first + 1;
+
+  auto one_shot_timer_callback = [this, timer_id]() mutable {
+    execute_check_expired_goals();
+
+    std::lock_guard<std::mutex> lock_mutex(expire_goal_timers_mutex_);
+    auto it = std::find_if(
+      expire_goal_timers_.begin(), expire_goal_timers_.end(),
+      [timer_id](const std::pair<unsigned int, rclcpp::TimerBase::SharedPtr>& element) {
+        return element.first == timer_id;
+      });
+
+    if (it != expire_goal_timers_.end()) {
+      expire_goal_timers_.erase(it);
+    }
+  };
+
+  auto timer = rclcpp::create_wall_timer(
+    std::chrono::nanoseconds(goal_expire_timeout_),
+    std::function<void()>(one_shot_timer_callback),
+    nullptr,
+    node_base_.get(),
+    node_timers_.get()
+  );
```

Comment:
just an idea. instead of constructing / destructing timer object for per goal result, can we keep the timer object and recalculate the timeout based on the current caching results? that is what the server does with wait set based executors.

Reply:
Done in f4408aa2ddaf1613b583e2602e99fdefbab8df6f